modules/ppcp-button/services.php

```diff
@@ -65,25 +67,86 @@
 
 		return $dummy_ids[ $shop_country ] ?? $container->get( 'button.client_id' );
 	},
-	'button.smart-button'                         => static function ( ContainerInterface $container ): SmartButtonInterface {
+	'button.is_paypal_continuation'               => static function( ContainerInterface $container ): bool {
+		$session_handler = $container->get( 'session.handler' );
+
+		$order = $session_handler->order();
+		if ( ! $order ) {
+			return false;
+		}
+		$source = $order->payment_source();
+		if ( $source && $source->card() ) {
+			return false; // Ignore for DCC.
+		}
+		if ( 'card' === $session_handler->funding_source() ) {
+			return false; // Ignore for card buttons.
+		}
+
+		return true;
+	},
+	'button.context'                              => static function( ContainerInterface $container ): string {
+		$is_paypal_continuation = $container->get( 'button.is_paypal_continuation' );
 
+		$context = 'mini-cart';
+		if ( is_product() || wc_post_content_has_shortcode( 'product_page' ) ) {
+			$context = 'product';
+		}
+		if ( is_cart() ) {
+			$context = 'cart';
+		}
+		if ( is_checkout() && ! $is_paypal_continuation ) {
+			$context = 'checkout';
+		}
+		if ( is_checkout_pay_page() ) {
+			$context = 'pay-now';
+		}
+
+		return $context;
+	},
+	'button.intent'                               => static function( ContainerInterface $container ): string {
+		$settings           = $container->get( 'wcgateway.settings' );
+		$subscription_helper = $container->get( 'subscription.helper' );
+		assert( $subscription_helper instanceof SubscriptionHelper );
+		$context = $container->get( 'button.context' );
+
+		$intent               = ( $settings->has( 'intent' ) ) ? $settings->get( 'intent' ) : 'capture';
+		$product_intent       = $subscription_helper->current_product_is_subscription() ? 'authorize' : $intent;
+		$other_context_intent = $subscription_helper->cart_contains_subscription() ? 'authorize' : $intent;
```

Comment:
Seems like `button.intent` is not used anywhere except passing to `SmartButton`, so it can be moved into a function inside `SmartButton` or somewhere else. Similarly to context, it is easy to accidentally misuse such services, call when not supposed to or too early.